python_onnxifi/data_conversion.cc

```diff
@@ -0,0 +1,251 @@
+    break;									 \
+  } 										 \
+  case ONNX_NAMESPACE::TensorProto_DataType_INT8:  {				 \
+    op_template<int32_t, int8_t>(__VA_ARGS__);                     		 \
+    break;									 \
+  } 	   	      								 \
+  case ONNX_NAMESPACE::TensorProto_DataType_INT16:  {				 \
+    op_template<int32_t, int16_t>(__VA_ARGS__);          			 \
+    break;									 \
+  } 										 \
+  case ONNX_NAMESPACE::TensorProto_DataType_INT32:  { 				 \
+    op_template<int32_t, int32_t>(__VA_ARGS__);          			 \
+    break;									 \
+  } 			       							 \
+  case ONNX_NAMESPACE::TensorProto_DataType_UINT8:  {				 \
+    op_template<int32_t, uint8_t>(__VA_ARGS__);          			 \
+    break;									 \
+  } 		 								 \
+  case ONNX_NAMESPACE::TensorProto_DataType_UINT16: { 				 \
+    op_template<int32_t, uint16_t>(__VA_ARGS__);          		    	 \
+    break;									 \
+  } 									         \
+  case ONNX_NAMESPACE::TensorProto_DataType_INT64: {				 \
+    op_template<int64_t, int64_t>(__VA_ARGS__);                               	 \
+    break;                                                               	 \
+  }										 \
+  case ONNX_NAMESPACE::TensorProto_DataType_UINT32:  {				 \
+    op_template<uint64_t, uint32_t>(__VA_ARGS__);                                \
+    break;                                                                       \
+  }     									 \
+  case ONNX_NAMESPACE::TensorProto_DataType_UINT64: {			 	 \
+    op_template<uint64_t, uint64_t>( __VA_ARGS__);                            	 \
+    break;							        	 \
+  }   						                        	 \
+  case ONNX_NAMESPACE::TensorProto_DataType_DOUBLE: {                   	 \
+   op_template<double, double>(__VA_ARGS__);                                 	 \
+   break;                                                                        \
+  }										 \
+  case ONNX_NAMESPACE::TensorProto_DataType_COMPLEX128: {     			 \
+   op_template<std::complex<double>, std::complex<double>>(__VA_ARGS__);         \
+   break;									 \
+  }										 \
+  case ONNX_NAMESPACE::TensorProto_DataType_STRING: 				 \
+  case ONNX_NAMESPACE::TensorProto_DataType_UNDEFINED: {			 \
+   throw std::runtime_error("Dispatch received non-numeric data type");          \
+  }										 \
+}										 \
+
+template<typename onnx_type, typename unused>
+void DataConversion::getBufferSize(uint64_t& buffer_size, const uint64_t* shape, uint32_t dimensions)  {
+  auto numElements = std::accumulate(shape, shape + dimensions,
+                                    (uint64_t) 1, std::multiplies<uint64_t>());
+  buffer_size = sizeof(onnx_type) * numElements;
+}
+
+void DataConversion::updateOutputDescriptors(ModelProto& model)  {
+  releaseDescriptors(output_descriptors_);
+  ::ONNX_NAMESPACE::shape_inference::InferShapes(model);
+  auto& outputs = model.graph().output();
+  for (auto& vi : outputs)  {
+    onnxTensorDescriptor outputDescriptor;
+    char* name = new char[vi.name().size() + 1];
```

Comment:
leak?

Reply:
releaseDescriptors handles the release of name, char, and buffer resources that are allocate
- For member variables, no need to actively call this as it is handled by the destructor and update functions
- I updated the header file to note that when directly using makeDescriptorsFromNumpy the release function must be called (these are availabe to use publicly for really only internal testing)
- I fixed a leak in my test to releaseDescriptors